File: android/src/main/java/com/zj/zjsdk/plugin/ZjNativeExpressAdView.java
package com.zj.zjsdk.plugin;

import android.app.Activity;
import android.content.Context;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;

import com.zj.zjsdk.ad.ZjAdError;
import com.zj.zjsdk.ad.ZjNativeExpressAd;
import com.zj.zjsdk.ad.ZjNativeExpressAdListener;
import com.zj.zjsdk.ad.ZjSize;

import java.util.HashMap;
import java.util.Map;

import io.flutter.plugin.common.BinaryMessenger;
import io.flutter.plugin.common.EventChannel;
import io.flutter.plugin.platform.PlatformView;

public class ZjNativeExpressAdView implements PlatformView {

    private static final String TAG = "ZjNativeExpressAdView";

    private final ViewGroup container;
    private EventChannel.EventSink mEventSink;

    public ZjNativeExpressAdView(Context context, Activity activity, BinaryMessenger messenger, int id, String posId, int width, int height) {

        EventChannel eventChannel = new EventChannel(messenger, "flutter_zjsdk_plugin/native_express_event_" + id);
        eventChannel.setStreamHandler(new EventChannel.StreamHandler() {
            @Override
            public void onListen(Object o, EventChannel.EventSink eventSink) {
                mEventSink = eventSink;
            }

            @Override
            public void onCancel(Object o) {
            }
        });

        container = new FrameLayout(context);
        container.setLayoutParams(new FrameLayout.LayoutParams(ZjUITool.toPx(context, width), ZjUITool.toPx(context, height)));

        loadAd(activity, posId, width, height);
    }

    private void loadAd(Activity activity, String posId, int width, int height) {

        ZjNativeExpressAd expressAd = new ZjNativeExpressAd(activity, posId, new ZjNativeExpressAdListener() {
            @Override
            public void onZjAdLoaded() {
                Map<String, Object> result = new HashMap<>();
                result.put("event", "onZjAdLoaded");
                mEventSink.success(result);
            }

            @Override
            public void onZjAdShow() {
                Map<String, Object> result = new HashMap<>();
                result.put("event", "onZjAdShow");
                mEventSink.success(result);
            }

            @Override
            public void onZjAdClicked() {
                Map<String, Object> result = new HashMap<>();
                result.put("event", "onZjAdClicked");
                mEventSink.success(result);
            }

            @Override
            public void onZjAdClosed() {
                Map<String, Object> result = new HashMap<>();
                result.put("event", "onZjAdClosed");
                mEventSink.success(result);
                mEventSink.endOfStream();
            }

            @Override
            public void onZjAdError(ZjAdError zjAdError) {
                Map<String, Object> result = new HashMap<>();
                result.put("event", "onZjAdError");
                result.put("code", zjAdError.getErrorCode());
                result.put("message", zjAdError.getErrorMsg());
                mEventSink.success(result);
                mEventSink.endOfStream();
            }
        }, container);
        expressAd.setSize(new ZjSize(width, height));
        expressAd.loadAd();
    }

    @Override
    public View getView() {
        return container;
    }

    @Override
    public void dispose() {
        Log.e(TAG, "dispose");
    }

}

File: lib/zj_native_express_ad_view.dart
import 'package:flutter/foundation.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

import 'zj_callback.dart';

/// Widget for ZjNativeExpressAd
class ZjNativeExpressAdView extends StatelessWidget {
  /// 广告位ID
  final String posId;

  /// 宽
  final double width;

  /// 高
  final double height;

  /// 广告加载成功
  final AdCallback0 onZjAdLoaded;

  /// 广告展示
  final AdCallback0 onZjAdShow;

  /// 广告点击
  final AdCallback0 onZjAdClicked;

  /// 广告关闭
  final AdCallback0 onZjAdClosed;

  /// 广告错误
  final AdErrorCallback onZjAdError;

  ZjNativeExpressAdView(
      {Key key,
      @required this.posId,
      @required this.width,
      @required this.height,
      this.onZjAdLoaded,
      this.onZjAdShow,
      this.onZjAdClicked,
      this.onZjAdClosed,
      this.onZjAdError})
      : super(key: key);

  @override
  Widget build(BuildContext context) {
    Widget nativeExpress;
    if (defaultTargetPlatform == TargetPlatform.android) {
      nativeExpress = AndroidView(
        viewType: 'flutter_zjsdk_plugin/native_express',
        creationParams: {
          "posId": posId,
          "width": width,
          "height": height,
        },
        creationParamsCodec: const StandardMessageCodec(),
        onPlatformViewCreated: _onPlatformViewCreated,
      );
    } else {
      nativeExpress = Text("Not supported");
    }

    return SafeArea(
      child: Stack(
        children: <Widget>[
          Container(
            width: width,
            height: height,
            child: nativeExpress,
          ),
        ],
      ),
    );
  }

  void _onPlatformViewCreated(int id) {
    EventChannel eventChannel =
        EventChannel("flutter_zjsdk_plugin/native_express_event_$id");
    eventChannel.receiveBroadcastStream().listen((event) {
      switch (event["event"]) {
        case "onZjAdLoaded":
          print("ZjNativeExpressAd.onZjAdLoaded");
          onZjAdLoaded?.call();
          break;

        case "onZjAdShow":
          print("ZjNativeExpressAd.onZjAdShow");
          onZjAdShow?.call();
          break;

        case "onZjAdClicked":
          print("ZjNativeExpressAd.onZjAdClicked");
          onZjAdClicked?.call();
          break;

        case "onZjAdClosed":
          print("ZjNativeExpressAd.onZjAdClosed");
          onZjAdClosed?.call();
          break;

        case "onZjAdError":
          print(
              "ZjNativeExpressAd.onZjAdError[${event["code"]}-${event["message"]}");
          onZjAdError?.call(event["code"], event["message"]);
          break;
      }
    });
  }
}
